ibis/expr/datatypes.py

```diff
@@ -1220,19 +962,317 @@ def type(self):
+    if castable(left, right, upcast=True):
+        return right
+    elif castable(right, left, upcast=True):
+        return left
+
+    raise com.IbisTypeError('Cannot compute precedence for {} '
+                            'and {} types'.format(left, right))
+
+
+def highest_precedence(dtypes):
+    return functools.reduce(higher_precedence, dtypes)
+
+
+@infer.register(object)
+def infer_dtype_default(value):
+    raise com.InputTypeError(value)
+
+
+@infer.register(OrderedDict)
+def infer_struct(value):
+    if not value:
+        raise TypeError('Empty struct type not supported')
+    return Struct(
+        list(value.keys()),
+        list(map(infer, value.values()))
+    )
+
+
+@infer.register(dict)
+def infer_map(value):
+    if not value:
+        return Map(null, null)
+    return Map(
+        highest_precedence(map(infer, value.keys())),
+        highest_precedence(map(infer, value.values())),
+    )
+
+
+@infer.register(list)
+def infer_list(value):
+    if not value:
+        return Array(null)
+    return Array(highest_precedence(map(infer, value)))
+
+
+@infer.register(np.ndarray)
+def infer_array(value):
+    # TODO: infer series
+    return Array(dtype(value.dtype.name))
+
+
+@infer.register(datetime.time)
+def infer_time(value):
+    return time
+
+
+@infer.register(datetime.date)
+def infer_date(value):
+    return date
+
+
+@infer.register(datetime.datetime)
+def infer_timestamp(value):
+    return timestamp
+
+
+@infer.register(datetime.timedelta)
+def infer_interval(value):
+    return interval
+
+
+@infer.register(six.string_types)
+def infer_string(value):
+    return string
+
+
+@infer.register(builtins.float)
+def infer_floating(value):
+    return double
+
+
+@infer.register(six.integer_types)
+def infer_integer(value, allow_overflow=False):
+    for dtype in [int8, int16, int32, int64]:
+        if dtype.bounds.lower <= value <= dtype.bounds.upper:
+            return dtype
+
+    if not allow_overflow:
+        raise OverflowError(value)
+
+    return int64
+
+
+@infer.register(np.generic)
+def infer_numpy_scalar(value):
+    return dtype(value.dtype)
+
+
+@infer.register(pd.Timestamp)
+def infer_pandas_timestamp(value):
+    return Timestamp(timezone=str(value.tz))
+
+
+@infer.register(bool)
+def infer_boolean(value):
+    return boolean
+
+
+@infer.register((type(None), Null))
+def infer_null(value):
+    return null
+
+
+castable = Dispatcher('castable')
+
+
+@castable.register(DataType, DataType)
+def can_cast_subtype(source, target, **kwargs):
+    return isinstance(target, type(source))
+
+
+@castable.register(Any, DataType)
+def can_cast_any(source, target, **kwargs):
+    return True
+
+
+@castable.register(Null, DataType)
+def can_cast_null(source, target, **kwargs):
+    return target.nullable
+
+
+@castable.register(Integer, Integer)
+def can_cast_to_generic_integer(source, target, **kwargs):
+    return True
+
+
+@castable.register(UnsignedInteger, UnsignedInteger)
```

Comment:
You can chain decorators here so you don't have to repeat the code for signed and unsigned integers, like this:

```python
@castable.register(UnsignedInteger, UnsignedInteger)
@castable.register(SignedInteger, SignedInteger)
```

Reply:
In fact, any of these implementations that take the same number of arguments and return `True` can all be chained into one function to avoid repetition.

Reply:
Cool!